File: backend/src/main/java/org/rdfarchitect/models/cim/data/dto/relations/uri/URI.java
/*
 *    Copyright (c) 2024-2026 SOPTIM AG
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package org.rdfarchitect.models.cim.data.dto.relations.uri;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import io.swagger.v3.oas.annotations.media.Schema;

import lombok.Getter;
import lombok.NoArgsConstructor;

import org.apache.jena.atlas.lib.StrUtils;
import org.apache.jena.graph.Node;
import org.apache.jena.graph.NodeFactory;
import org.apache.jena.irix.IRIException;
import org.apache.jena.irix.IRIx;

@NoArgsConstructor
@JsonDeserialize(using = URIDeserializer.class)
@Schema(
        description =
                "When deserializing (parsing json to object), it is possible to alternatively only put a String containing the whole uri. For serializing (parsing object "
                        + "to json), the object structure ist always used.")
public class URI {

    @Getter private String prefix;
    @Getter private String suffix;

    public URI(String uri) {
        if (uri == null) {
            throw new IllegalArgumentException("URI must not be null");
        }
        if (uri.isEmpty()) {
            throw new IllegalArgumentException("URI must not be empty");
        }
        if (uri.contains(" ")) {
            throw new IllegalArgumentException("URI must not contain spaces: " + uri);
        }
        try {
            IRIx iri = IRIx.create(StrUtils.encodeHex(uri, '%', new char[] {'%', ' '}));
            if (iri.scheme() == null) {
                throw new IllegalArgumentException("IRI must be absolute: " + uri);
            }
        } catch (IRIException e) {
            throw new IllegalArgumentException("Invalid IRI: " + uri, e);
        }
        int hash = uri.indexOf('#');
        int slash = uri.lastIndexOf('/');

        if (hash >= 0) {
            this.prefix = uri.substring(0, hash + 1);
            this.suffix = uri.substring(hash + 1);
        } else if (slash >= 0) {
            this.prefix = uri.substring(0, slash + 1);
            this.suffix = uri.substring(slash + 1);
        } else {
            this.prefix = uri;
            this.suffix = "";
        }
    }

    @Override
    public String toString() {
        return prefix + suffix;
    }

    public Node toNode() {
        return NodeFactory.createURI(this.toString());
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof URI other) {
            return this.toString().equals(other.toString());
        }
        return false;
    }

    @Override
    public int hashCode() {
        return this.toString().hashCode();
    }
}

File: frontend/src/lib/models/dto/uri.ts
/*
 *    Copyright (c) 2024-2026 SOPTIM AG
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

import { validateIri } from "validate-iri";

export class URI {
    prefix: string | null = null;
    suffix: string = "";

    constructor(input: string | { prefix: string | null; suffix: string }) {
        if (typeof input === "string") {
            const { prefix, suffix } = URI.parse(input);
            this.prefix = prefix;
            this.suffix = suffix;
        } else {
            this.prefix = input.prefix ?? null;
            this.suffix = input.suffix ?? "";
        }
    }

    private static parse(input: string): {
        prefix: string | null;
        suffix: string;
    } {
        if (!input) {
            throw new Error("URI must not be null or empty");
        }

        const error = validateIri(input);
        if (error) {
            throw new Error(`Invalid IRI: ${input}`);
        }

        const hash = input.indexOf("#");
        const slash = input.lastIndexOf("/");

        let prefix: string | null;
        let suffix: string;
        if (hash >= 0) {
            prefix = input.substring(0, hash + 1);
            suffix = input.substring(hash + 1);
        } else if (slash >= 0) {
            prefix = input.substring(0, slash + 1);
            suffix = input.substring(slash + 1);
        } else {
            prefix = null;
            suffix = input;
        }

        if (!prefix) {
            throw new Error(`IRI must have a namespace: ${input}`);
        }

        return { prefix, suffix };
    }

    toString(): string {
        return (this.prefix ?? "") + this.suffix;
    }

    equals(other: URI): boolean {
        if (!(other instanceof URI)) {
            return false;
        }
        return this.toString() === other.toString();
    }
}
